File: bot/src/spacetraders_bot/core/system_graph_provider.py
"""Shared service for loading and caching system navigation graphs."""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import Dict, Optional

from .database import Database
from .route_planner import GraphBuilder

logger = logging.getLogger(__name__)


@dataclass
class GraphLoadResult:
    """Details about a system graph load operation."""

    graph: Dict
    source: str
    message: str | None = None


class SystemGraphProvider:
    """Provide system graphs with consistent caching and persistence."""

    def __init__(
        self,
        api_client,
        *,
        db: Optional[Database] = None,
        builder: Optional[GraphBuilder] = None,
    ) -> None:
        self.api = api_client
        self.db = db or Database()
        self._builder = builder or GraphBuilder(api_client, db_path=self.db.db_path)

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def get_graph(self, system_symbol: str, *, force_refresh: bool = False) -> GraphLoadResult:
        """Return the navigation graph for *system_symbol*."""
        if not force_refresh:
            graph = self._load_from_database(system_symbol)
            if graph is not None:
                return GraphLoadResult(
                    graph=graph,
                    source="database",
                    message=f"📊 Loaded graph for {system_symbol} from database",
                )

        graph = self._build_from_api(system_symbol)
        return GraphLoadResult(
            graph=graph,
            source="api",
            message=f"📊 Built graph for {system_symbol} from API",
        )

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    def _load_from_database(self, system_symbol: str) -> Optional[Dict]:
        with self.db.connection() as conn:
            graph = self.db.get_system_graph(conn, system_symbol)
            if graph is not None:
                logger.info("Loaded %s graph from database", system_symbol)
            return graph

    def _build_from_api(self, system_symbol: str) -> Dict:
        """Build graph from API and save to database."""
        logger.info("Building navigation graph for %s from API", system_symbol)
        graph = self._builder.build_system_graph(system_symbol)
        if not graph:
            raise RuntimeError(f"Failed to build graph for {system_symbol}")
        # Note: GraphBuilder.build_system_graph() already saves to database
        logger.info("Graph for %s saved to database", system_symbol)
        return graph

File: gobot/internal/adapters/graph/system_graph_provider.go
package graph

import (
	"context"
	"fmt"
	"log"

	"github.com/andrescamacho/spacetraders-go/internal/domain/system"
)

// SystemGraphProvider provides system navigation graphs with database caching
//
// Checks database cache first, falls back to building from API if needed.
// Stores newly built graphs in database for future use.
type SystemGraphProvider struct {
	graphRepo    system.SystemGraphRepository
	graphBuilder system.IGraphBuilder
	playerID     int
}

// NewSystemGraphProvider creates a new system graph provider
func NewSystemGraphProvider(
	graphRepo system.SystemGraphRepository,
	graphBuilder system.IGraphBuilder,
	playerID int,
) system.ISystemGraphProvider {
	return &SystemGraphProvider{
		graphRepo:    graphRepo,
		graphBuilder: graphBuilder,
		playerID:     playerID,
	}
}

// GetGraph retrieves system navigation graph (checks cache first, builds from API if needed)
func (p *SystemGraphProvider) GetGraph(ctx context.Context, systemSymbol string, forceRefresh bool) (*system.GraphLoadResult, error) {
	// Try loading from database cache first (unless force refresh)
	if !forceRefresh {
		graph, err := p.loadFromDatabase(ctx, systemSymbol)
		if err != nil {
			log.Printf("Error loading graph from database: %v", err)
		} else if graph != nil {
			log.Printf("Loaded graph for %s from database cache", systemSymbol)
			return &system.GraphLoadResult{
				Graph:   graph,
				Source:  "database",
				Message: fmt.Sprintf("Loaded graph for %s from database cache", systemSymbol),
			}, nil
		}
	}

	// Build from API and cache it
	graph, err := p.buildFromAPI(ctx, systemSymbol)
	if err != nil {
		return nil, err
	}

	return &system.GraphLoadResult{
		Graph:   graph,
		Source:  "api",
		Message: fmt.Sprintf("Built graph for %s from API", systemSymbol),
	}, nil
}

// loadFromDatabase loads graph from database cache
func (p *SystemGraphProvider) loadFromDatabase(ctx context.Context, systemSymbol string) (map[string]interface{}, error) {
	graph, err := p.graphRepo.Get(ctx, systemSymbol)
	if err != nil {
		return nil, fmt.Errorf("failed to load graph: %w", err)
	}

	if graph != nil {
		log.Printf("Cache hit for %s", systemSymbol)
	} else {
		log.Printf("Cache miss for %s", systemSymbol)
	}

	return graph, nil
}

// buildFromAPI builds graph from API and saves to database
func (p *SystemGraphProvider) buildFromAPI(ctx context.Context, systemSymbol string) (map[string]interface{}, error) {
	log.Printf("Building navigation graph for %s from API", systemSymbol)

	// Build the graph using this player's API client
	graph, err := p.graphBuilder.BuildSystemGraph(ctx, systemSymbol, p.playerID)
	if err != nil {
		return nil, fmt.Errorf("failed to build graph for %s: %w", systemSymbol, err)
	}

	// Save to database cache
	if err := p.saveToDatabase(ctx, systemSymbol, graph); err != nil {
		log.Printf("Warning: failed to save graph to database: %v", err)
		// Don't fail - caching failure shouldn't break the operation
	} else {
		log.Printf("Graph for %s cached in database", systemSymbol)
	}

	return graph, nil
}

// saveToDatabase saves graph to database cache
func (p *SystemGraphProvider) saveToDatabase(ctx context.Context, systemSymbol string, graph map[string]interface{}) error {
	if err := p.graphRepo.Save(ctx, systemSymbol, graph); err != nil {
		return fmt.Errorf("failed to save graph: %w", err)
	}

	log.Printf("Saved graph for %s to database", systemSymbol)
	return nil
}
